kernel/sched.c

```diff
@@ -735,16 +762,25 @@ static void z_thread_halt(struct k_thread *thread, k_spinlock_key_t key,
 
 			/* Now we know it's halting, but not necessarily
 			 * halted (suspended or aborted). Wait for the switch
-			 * to happen!
+			 * to happen unless deadlock was detected. If deadlock
+			 * was detected, we know we are in a special case where
+			 * all relevant CPUs are in ISRs trying to halt one or
+			 * more threads and no context switch can be done
+			 * until one of the ISRs exit.
```

Comment:
I'm not completely understanding this paragraph.  How about "If we detect a deadlock, then don't spin.  We will be the first CPU of the chain of spinners to complete, and the others will exit their spin loops in turn."

Reply:
That works. It shall be so.